### PR TITLE
Added PASpatialPooler, PALayer, and adjusted Layer.java for inheritance

### LINKED ISSUE #340: Change two private methods to protected in Layer.java
I'm looking at adding paCLA to htm.java using as little new code as possible. I want to inherit from Layer, SpatialPooler and TemporalMemory. In Layer, I just need to override spatialInput() and temporalInput(), but they're marked as private.

### PR DESCRIPTION
**Pull request by fergalbyrne**
Fixes #340

### CODE REVIEW

**Comment by cogmission** [comment]
This is awesome @fergalbyrne ! :+1: 

Coveralls is complaining because their were no tests added for the new code. If you like I can add some new tests, if you can describe to me the general overall behavior difference when using PA-awesomeness, like TM converging faster or something? But in a little bit more detail so I can write tests for this? In that case I will go ahead and merge this and add tests myself later tonight to appease Coveralls.

Either that, or if you want to add the tests that is fine also - and I will delay merging until you are ready?

Let me know either way?

**Comment by cogmission** [comment]
Eventually I would like to add a comparison module that illustrates HTM performance gains when using paCLA.

**Comment by fergalbyrne** [comment]
I'll do the tests. Want to add setters for the mixing parameter, so we can turn the PA up and down. Do you have tests using the Network API? We should be able to use that for tests and comparison.

**Comment by cogmission** [comment]
the src/test/java dir has tests according to package. There are several in the ```org.nument.nupic.network``` package, namely: LayerTest.java; RegionTest.java; and NetworkTest.java.

One thing to note is there is a convention in HTM.java where the algorithms hold no state. Specialized parameters are stored in the Parameters.java class and state is stored in the Connections.java class. The ```calculateOverlap()``` method already passes in the Connections object so all you have to do is store ```protected int[] paOverlaps;``` in there and create a setter/getter pair for it.

In summary, depending on the use case, state or parameters are stored in either: Parameters (for configuration), Connections (for actual state), ManualInput/Inference (for passing data between algorithms and layers). 

The Layer class does in fact hold state so it is ok to duplicate often used fields there for user convenience because its often easier for user to just call network.getXXX() or layer..getXXX() for the most recent values in the current compute cycle.

**Comment by cogmission** [comment]
Hi Fergal,

Excellent Work! :+1: 

Can you remove that line from the PASpatialPooler?
https://github.com/fergalbyrne/htm.java/blob/paCLA/src/main/java/org/numenta/nupic/algorithms/PASpatialPooler.java#L48

**Comment by cogmission** [comment]
After you do the above we'll merge... :+1:

**Comment by fergalbyrne** [comment]
PALayer is a Layer which can contain a PASpatialPooler, in which case (via Connections) it prepopulates the paSP's overlap vector with a depolarisation value derived from its TM's predictive cells. PASpatialPooler adds this vector to the overlaps calculated from the feedforward input before doing inhibition. This change pre-biases the paSP to favour columns with predictive cells. Full details at http://arxiv.org/abs/1509.08255

**Comment by cogmission** [comment]
That's good for the class documentation, but I was looking for... 
```
When "paDepolarize" is set to "0.0" the polarization value is not added to the SpatialPooler's overlap.
When increasing this number it has X effect?
```
